### PR TITLE
Fix expired node in NodeStore not sync with SessionManager

### PR DESCRIPTION
**Pull request by madafoo**
Every 30 seconds, the SessionManage will sync the established nodes with the Pathfinder NodeStore nodes.
If the node in NodeStore is invalid state but it is established in SessionManage, the expired node will be destroyed from the NodeStore.

### CODE REVIEW

File: dht/Pathfinder.c [hunk 1]
```diff
@@ -363,6 +363,22 @@ static Iface_DEFUN handlePong(struct Message* msg, struct Pathfinder_pvt* pf)
     return NULL;
 }
 
+static Iface_DEFUN syncNode(struct Message* msg, struct Pathfinder_pvt* pf)
+{
+    uint8_t addr[16];
+    Message_pop(msg, addr, 16, NULL);
+    Assert_true(!msg->length);
+    {
```
Comment (cjdelisle) [inline dht/Pathfinder.c]: (informational)
```if (Defined(Log_DEBUG)) {``` will work here. Note: you need to ```#import "util/Defined"```
and don't ask how it works :)

File: dht/dhtcore/NodeStore.c [hunk 2]
```diff
@@ -2442,3 +2442,17 @@ uint64_t NodeStore_timeSinceLastPing(struct NodeStore* nodeStore, struct Node_Tw
     }
     return now - lastSeen;
 }
+
+void NodeStore_destroyNode(struct NodeStore* nodeStore, uint8_t* addr)
```
Comment (cjdelisle) [inline dht/dhtcore/NodeStore.c]: Name of this function is problematic, it indicates that the node is to be destroyed (unconditionally) but it's behaviour is NodeStore_destroyNodeIfNoKnownPath()

**Comment by cjdelisle** [comment]
This is a great PR, however I am not sure if this approach really solves the problem. Can you clarify for me why exactly kicking an unreachable node from the NodeStore will solve the issue in #804. It seems to me that the correct behaviour is:

If NodeStore.node becomes unreachable AND SessionManager contains node with *same path*, reset SessionManager path to "unknown". If this is correct then the codepath should follow something like:

https://github.com/cjdelisle/cjdns/blob/master/dht/dhtcore/NodeStore.c#L332
https://github.com/cjdelisle/cjdns/blob/master/dht/Pathfinder.c#L164
https://github.com/cjdelisle/cjdns/blob/master/dht/Pathfinder.c#L134
https://github.com/cjdelisle/cjdns/blob/master/net/SessionManager.c#L552

**Comment by madafoo** [comment]
In fact, I have seen many case is that NodeStore.node become unreachable but SessionManager.node path is changed for renumbered reason.
Many case I even call the SeachRunner_search from the admin api, the expired NodeStore.node still prevent the correct reponse of DHT findNodes to update NodeStore table.
Hence I have to find another way to flush/destroy the expired NodeStore.node

**Comment by cjdelisle** [comment]
Kicking a search in response to having your known-path demolished inside of the NodeStore sounds like a very good idea.

**Comment by madafoo** [comment]
So my first patch is launching a search  in response to get the response of findNodes. But I find the NodeStore.discoverLinkC will ignore the new path.

**Comment by cjdelisle** [comment]
Perhaps we should just add a triggerSearch() call to https://github.com/cjdelisle/cjdns/blob/master/net/SessionManager.c#L554 ?
It still carries some risk but it is in the right logical place for such a thing...

**Comment by cjdelisle** [comment]
I'm still not sure, given that code, why you are getting the bug behaviour -- shouldn't it dump the label from the session ?

**Comment by madafoo** [comment]
> https://github.com/cjdelisle/cjdns/blob/master/dht/dhtcore/NodeStore.c#L332
> https://github.com/cjdelisle/cjdns/blob/master/dht/Pathfinder.c#L164
> https://github.com/cjdelisle/cjdns/blob/master/dht/Pathfinder.c#L134
> https://github.com/cjdelisle/cjdns/blob/master/net/SessionManager.c#L552

Above is the codepath to handle when onBestPathChanged triggled by NodeStore accept the new path, but my problem is that under some case the NodeStore will not accept the new path for expired NodeStore prevent the new path to be accepted.

**Comment by madafoo** [comment]
For example, firstHopInPath in NodeStore will not function proper when the expired node exists. The link for this expired node in fact is invalid for a long time.

**Comment by cjdelisle** [comment]
Hmm, this might be the problem here:
https://github.com/cjdelisle/cjdns/blob/master/dht/Pathfinder.c#L387

What it is supposed to do is send a NODE message to the SessionManager every time a node sends you a message and SessionManager does not already know him. However it is possible that this if statement should be removed or changed and it should send the NODE message unconditionally...

WDYT?

**Comment by cjdelisle** [comment]
Note that when you kick a search for a node, this function incomingMsg() will be triggered when he responds which is also what completes the search.

**Comment by madafoo** [comment]
> https://github.com/cjdelisle/cjdns/blob/master/dht/Pathfinder.c#L387

Yeah, I think it should send the NODE message unconditionally.
A good place to sync two node table.

But it is not fix the expired node problem, I will capture more detail on that case.

**Comment by cjdelisle** [comment]
dropped the if statement from line 387, if you can get more test results to nail down the source of the problem, that would be excellent.

**Comment by madafoo** [comment]
More detail on NodeStore.discoverLinkC:
https://github.com/cjdelisle/cjdns/blob/master/dht/dhtcore/NodeStore.c#L996

findClosest return 1 means NodeStore links find the closest links, but
https://github.com/cjdelisle/cjdns/blob/master/dht/dhtcore/NodeStore.c#L1040

Why it think the matched closest link is self route? Do not understand why it use pathParentChild here?

**Comment by cjdelisle** [comment]
discussion continued in #811

**Pull request closed without merging**